From 49aab2a503a28a6eabe8e28c7fc7ade65ccca721 Mon Sep 17 00:00:00 2001
From: Pierre-Yves MORDRET <pierre-yves.mordret@st.com>
Date: Fri, 4 Oct 2019 09:47:43 +0200
Subject: [PATCH 13/14] waylandsink: XDG protocol does not work in fullscreen

Using new XDG protocol, it turns out to be not working when enabling
fullscreen support.
This patch avoids to over commit surface during configuration and takes
into account render region when provided within wayland callback.

Signed-off-by: Pierre-Yves MORDRET <pierre-yves.mordret@st.com>
---
 ext/wayland/wlwindow.c | 21 ++++++++++++++-------
 1 file changed, 14 insertions(+), 7 deletions(-)

diff --git a/ext/wayland/wlwindow.c b/ext/wayland/wlwindow.c
index 77b2d81..c29f307 100644
--- a/ext/wayland/wlwindow.c
+++ b/ext/wayland/wlwindow.c
@@ -253,7 +253,7 @@ gst_wl_window_new_toplevel (GstWlDisplay * display, const GstVideoInfo * info,
     gboolean fullscreen, GMutex * render_lock)
 {
   GstWlWindow *window;
-  gint width;
+  gint width, height;
 
   window = gst_wl_window_new_internal (display, render_lock);
 
@@ -311,10 +311,15 @@ gst_wl_window_new_toplevel (GstWlDisplay * display, const GstVideoInfo * info,
     goto error;
   }
 
-  /* set the initial size to be the same as the reported video size */
-  width =
-      gst_util_uint64_scale_int_round (info->width, info->par_n, info->par_d);
-  gst_wl_window_set_render_rectangle (window, 0, 0, width, info->height);
+  if ( window->render_rectangle.w == 0 || window->render_rectangle.h == 0) {
+    width =
+        gst_util_uint64_scale_int_round (info->width, info->par_n, info->par_d);
+    height = info->height;
+  } else {
+    width = window->render_rectangle.w;
+    height = window->render_rectangle.h;
+  }
+  gst_wl_window_set_render_rectangle (window, 0, 0, width, height);
 
   return window;
 
@@ -541,8 +546,10 @@ gst_wl_window_set_render_rectangle (GstWlWindow * window, gint x, gint y,
     gst_wl_window_resize_video_surface (window, TRUE);
   }
 
-  wl_surface_damage (window->area_surface_wrapper, 0, 0, w, h);
-  wl_surface_commit (window->area_surface_wrapper);
+  if (window->xdg_surface && window->configured) {
+    wl_surface_damage (window->area_surface_wrapper, 0, 0, w, h);
+    wl_surface_commit (window->area_surface_wrapper);
+  }
 
   if (window->video_width != 0)
     wl_subsurface_set_desync (window->video_subsurface);
-- 
2.7.4

